### DISCUSSION TITLE
What permissions should I use?

### DISCUSSION THREAD

**Discussion by justinmchase** [github/safe-settings#427] [Q&A] [answered]
The docs say to use one of the permisisons `admin`, `push`, `pull`.

But these don't align with with Roles in my orgs.

![Screen Shot 2023-03-02 at 5 05 56 PM](https://user-images.githubusercontent.com/10974/222582047-cbe0ed0f-1e87-4798-ae3c-417ab94f9d2d.png)

Is there any documentation on what Role will be applied if I use "push" in this case? Or should I be using permission names such as `maintain` instead of `push` or `pull` if those are the Roles defined in my org?

**Comment by svg153** [comment, marked as answer]
Hi. We are using `admin`, `push`, `pull` in our safe-settings configuration. So, I could confirm you that`pull` works as a `read`, and `pull` as a `write`. But we have not yet tested these "new roles" or the custom roles.

**Reply by justinmchase** [reply]
I can't tell that they are new or custom, since they seem to be set at a higher level than I have permissions for. So your roles in your org are literally `push` and `pull`? 

It actually seems to work to set `maintain`, knowing the code is there any reason to think it wouldn't work?

**Reply by svg153** [reply]
By `"new roles"` I meant the `triage` and `maintain` roles (These roles [were released in 2019](https://github.blog/changelog/2019-08-30-updates-to-triage-and-maintain-roles-beta/). Obviously they are not new, but they are different from the usuals)
We have, in the GitHub interface, the same roles as you. They are the default of organizations. But in settings we only use `push`, `pull` and `admin`.

**Reply by justinmchase** [reply]
Ok, not gonna lie, I'm confused. Sorry, not trying to be obtuse but if I use the `permission: maintain` will it not set the team correctly? 

Does the permission value just need to match the name of the role you wish to assign the team to, regardless of the organizations roles? Or is this a bug in this app where "push", "pull" and "admin" are the only roles supported?

**Reply by svg153** [reply]
Forgive me, I didn't explain myself well, either.

After looking at the [safe-settings code for teams](https://github.com/github/safe-settings/blob/b0aa116f13e40ed6f35704265af79d829e92c1b2/lib/plugins/teams.js#L26), and the [GitHub API for teams](https://docs.github.com/en/rest/teams/teams?apiVersion=2022-11-28#add-or-update-team-repository-permissions), it seems to only support this list of roles:

> The permission to grant the team on this repository. We accept the following permissions to be set: `pull`, `triage`, `push`, `maintain`, `admin`

It also supports custom roles created, but I don't know how to assign them through safe-settings… I don't know if putting the name of the role would work, or it would have to be the identifier.

So even though the read and write appear in the interface, they are not valid identifiers for the API. Looks like some UX.

**Reply by justinmchase** [reply]
Ok so you're saying Read and Write are just labels on the drop down but when it calls the api its actually sending "pull" and "push" and those are the actual valid settings settings bot is using.

What about custom roles? They also have identifiers?

**Reply by svg153** [reply]
After some tests:

- GitHub API (with new and legacy API)
- safe-settings in my org (use the old legacy API for teams)

I could confirm you, that safe-settings works for custom roles with the name, btw double quotes `permission: "Custom Role 3"`

For example:
- Create manually these teams in your organization: "test_custom_role", "test_custom_role_2", "test_custom_role_3"
- Create manually these roles in your organization: "customrole", "custom role 2", "Custom Role 3".
- In the repository (or `settings.yml`) add the team part as below, for example, for `api.yml` repository.

```yaml
repository:
  name: api
teams:
  - name: test_custom_role
    permission: "customrole"
  - name: test_custom_role_2
    permission: "rustom role 2"
  - name: test_custom_role_3
    permission: "Custom Role 3"
```

**Reply by justinmchase** [reply]
So the permission field should match the Role name. Thanks.
